src/main.rs

```diff
@@ -97,85 +258,50 @@ fn main() {
     let username = username.expect("<username> argument");
     let password = password.expect("<password> argument");
 
-    let mut target = TargetJson {
-        major_version: MAJOR_VERSION,
-        minor_version: MINOR_VERSION,
-        x: 0,
-        y: 0,
-        image: String::new()
-    };
-    let mut width = 0;
-    let mut height = 0;
-    let mut target_image = Vec::new();
+    let init_config = matches.value_of("config").unwrap_or(TARGET_JSON_URL);
 
     let mut board = Vec::new();
     board.resize(1000*1000/2, 0u8);
 
-    loop {
-        let mut try_place_pixel = || -> Result<(), Box<Error>> {
-            let new_target: TargetJson = reqwest::get(TARGET_JSON_URL)?.json()?;
-            if new_target.major_version > MAJOR_VERSION {
-                println!("New major version is available. Must update!");
-                process::exit(1);
-            }
+    let root = Rc::new(RefCell::new(Job::new(init_config.to_string())));
 
-            if new_target.minor_version > MINOR_VERSION {
-                println!("New minor version is available. Update when convenient.");
-            }
-
-            if new_target != target {
-                target = new_target;
-                let mut decoder = png::Decoder::new(reqwest::get(&target.image)?);
-                decoder.set(png::TRANSFORM_EXPAND | png::TRANSFORM_GRAY_TO_RGB | png::TRANSFORM_PACKING | png::TRANSFORM_STRIP_16);
-                let (info, mut reader) = decoder.read_info()?;
-                width = info.width;
-                height = info.height;
-                let mut buffer = Vec::new();
-                buffer.resize(info.buffer_size(), 0u8);
-                reader.next_frame(&mut *buffer)?;
-
-                target_image.truncate(0);
-                target_image.reserve_exact((width*height) as usize);
-
-                match info.color_type {
-                    png::ColorType::RGB => {
-                        for color in buffer.chunks(3) {
-                            let c = [color[0], color[1], color[2], 255];
-                            target_image.push(color_to_index(&c));
-                        }
-                    },
-                    png::ColorType::RGBA => {
-                        for color in buffer.chunks(4) {
-                            target_image.push(color_to_index(color));
+    loop {
+        let mut queue = VecDeque::new();
+        queue.push_front(root.clone());
+
+        let mut already_seen = HashSet::new();
+
```

Comment:
The HashMap is for detecting cycles in the fallback recursion. This way, configs of different projects can link to each other without issues.